Split deserialize from runPythonUdf and remove generatePythonUDFResult

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34497 Don't run user function until all UserRRefs in the args are confirmed
* #34809 Fix Error Hanlding for Local Send
* **#34496 Split deserialize from runPythonUdf and remove generatePythonUDFResult**



Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

mrshenli committed:
Split deserialize from runPythonUdf and remove generatePythonUDFResult

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -70,11 +70,17 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+      auto& pythonRpcHandler = PythonRpcHandler::getInstance();
+      std::shared_ptr<SerializedPyObj> serializedPyObj = nullptr;
+      {
+        pybind11::gil_scoped_acquire ag;
```
do we need GIL here?  serialize and deserialize already hold a GIL inside

zhaojuanmao commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -140,9 +146,15 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+        pybind11::gil_scoped_acquire ag;
+        auto pythonUdf = pythonRpcHandler.deserialize(prc.serializedPyObj());
+        py_ivalue = jit::toIValue(
+            PythonRpcHandler::getInstance().runPythonUdf(std::move(pythonUdf)),
```
same as above, does it need GIL here?
pythonRpcHandler.runPythonUdf()

mrshenli replied:
When I drop the GIL, I hit errors/hangs in tests. Looks like even std::move an `py::object` requires GIL?

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Looks good, although CI says there was a failure introduced in this PR.

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

mrshenli committed:
Update on "Split deserialize from runPythonUdf and remove generatePythonUDFResult"




Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Stamp, although please make sure tests pass
